web-api/src/middleware/request_context.rs

```diff
@@ -0,0 +1,184 @@
+                String::from_utf8_lossy(value.as_bytes())
+            )
+        })?;
+        header
+            .parse::<Uuid>()
+            .map_err(|_| anyhow!("Non UUID {TENANT_ID_HEADER} header: {header}"))
+            .map(TenantId)
+    }
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display, Serialize)]
+#[serde(transparent)]
+pub(crate) struct RequestId(Uuid);
+
+impl RequestId {
+    pub(crate) fn generate() -> Self {
+        Self(Uuid::new_v4())
+    }
+
+    pub(crate) const fn missing() -> Self {
+        Self(Uuid::nil())
+    }
+}
+
+/// Setups the call context.
+///
+/// This make the `RequestId` and `TenantId` available as extensions and sets up tracing for all calls.
+///
+/// The `TenantId` is seen as _required_, i.e. the middleware will fail any
+/// request where it can't find a tenant id.
+pub(crate) fn setup_request_context<A, S>(
+    request: ServiceRequest,
+    service: &S,
+) -> impl Future<Output = Result<ServiceResponse<BoxBody>, actix_web::Error>> + 'static
+where
+    A: AsRef<tenants::Config> + 'static,
+    S: Service<ServiceRequest, Response = ServiceResponse<BoxBody>, Error = actix_web::Error>,
+    S::Future: 'static,
+{
+    let request_id = RequestId::generate();
+    let config = request.app_data::<Data<A>>().unwrap().get_ref().as_ref();
+
+    let tenant_id = match extract_tenant_id(config, &request) {
+        Ok(id) => id,
+        Err(error) => {
+            let response = middleware_failure(
+                "setup_request_context",
+                request,
+                Some(request_id),
+                None,
+                error,
+            );
+            return Either::Left(future::ok(response));
+        }
+    };
+
+    let span = info_span!(
+        "request",
+        path = %request.request().path(),
+        method = %request.request().method(),
+        request_id = %request_id,
+        tenant_id = %tenant_id,
+    );
+
+    trace!(parent: &span, "request received");
+
+    let context = Arc::new(RequestContext {
+        tenant_id,
+        request_id,
+    });
+
+    request.extensions_mut().insert(context);
+
+    Either::Right(
+        service
+            .call(request)
+            .instrument(span.clone())
+            .inspect(|_| trace!(parent: span, "request processed")),
+    )
+}
+
+const TENANT_ID_HEADER: &str = "X-Tenant-Id";
+
+fn extract_tenant_id(
+    config: &tenants::Config,
+    request: &ServiceRequest,
+) -> Result<TenantId, anyhow::Error> {
+    let header_value = request
+        .headers()
+        .get(TENANT_ID_HEADER)
+        .map(TenantId::try_from)
+        .transpose()?;
+
+    match header_value {
+        //FIXME in follow up PR this ID will be fetched from the database
+        //      during startup/storage initialization.
+        None if config.enable_legacy_tenant => Ok(TenantId::missing()),
+        None => Err(anyhow!("{TENANT_ID_HEADER} header missing")),
+        Some(passed_value) => Ok(passed_value),
+    }
```

Comment:
:wrench: We don't have to check the header if we are not multi-tenant
```suggestion
  if config.enable_legacy_tenant {
        return Ok(TenantId::missing());
    }

    request
        .headers()
        .get(TENANT_ID_HEADER)
        .map(TenantId::try_from)
        .ok_or_else(||  anyhow!("{TENANT_ID_HEADER} header missing"))?
```

Reply:
we do like mentioned above there is no multi tenant mode, just legacy tenants which will be used by default if no tenant id is given explicitly and the legacy tenant is enabled.

Reply:
So we can have the header and `enable_legacy_tenant= true`? That seems an error.

Reply:
yes, means less hassle when migrating much later down the line and forbidding it is additional work with doesn't help us in any way.

So less work and less problems further down the line seems like a win-win for me.